RecoHI/HiJetAlgos/plugins/HiPuRhoProducer.cc

```diff
@@ -0,0 +1,645 @@
+
+  for (std::size_t i = 0; i < size; ++i) {
+    auto const& index = order[i].first;
+
+    sortedEtaEdgeLow[i] = etaEdgeLow_[index];
+    sortedEtaEdgeHigh[i] = etaEdgeHi_[index];
+
+    (*mapToRhoOut)[i] = rho_[index];
+    (*mapToRhoExtraOut)[i] = rhoExtra_[index];
+    (*mapToRhoMOut)[i] = rhoM_[index];
+    (*mapToNTowOut)[i] = nTow_[index];
+  }
+
+  auto mapToRhoMedianOut = std::make_unique<std::vector<double>>(size);
+
+  for (uint32_t index = medianWindowWidth_; index < size - medianWindowWidth_; ++index) {
+    auto centre = mapToRhoOut->begin() + index;
+    std::vector<float> rhoRange(centre - medianWindowWidth_, centre + medianWindowWidth_);
+    std::nth_element(rhoRange.begin(), rhoRange.begin() + medianWindowWidth_, rhoRange.end());
+    (*mapToRhoMedianOut)[index] = rhoRange[medianWindowWidth_];
+  }
+
+  auto mapEtaRangesOut = std::make_unique<std::vector<double>>();
+
+  mapEtaRangesOut->assign(sortedEtaEdgeLow.begin(), sortedEtaEdgeLow.end());
+  mapEtaRangesOut->push_back(sortedEtaEdgeHigh.back());
+
+  auto mapToTowExcludePtOut = std::make_unique<std::vector<double>>(std::move(towExcludePt_));
+  auto mapToTowExcludePhiOut = std::make_unique<std::vector<double>>(std::move(towExcludePhi_));
+  auto mapToTowExcludeEtaOut = std::make_unique<std::vector<double>>(std::move(towExcludeEta_));
+
+  iEvent.put(std::move(mapEtaRangesOut), "mapEtaEdges");
+  iEvent.put(std::move(mapToRhoOut), "mapToRho");
+  iEvent.put(std::move(mapToRhoMedianOut), "mapToRhoMedian");
+  iEvent.put(std::move(mapToRhoExtraOut), "mapToRhoExtra");
+  iEvent.put(std::move(mapToRhoMOut), "mapToRhoM");
+  iEvent.put(std::move(mapToNTowOut), "mapToNTow");
+  iEvent.put(std::move(mapToTowExcludePtOut), "mapToTowExcludePt");
+  iEvent.put(std::move(mapToTowExcludePhiOut), "mapToTowExcludePhi");
+  iEvent.put(std::move(mapToTowExcludeEtaOut), "mapToTowExcludeEta");
+}
+
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+void HiPuRhoProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  //The following says we do not know what parameters are allowed so do no validation
+  // Please change this to state exactly what you do use, even if it is no parameters
+  edm::ParameterSetDescription desc;
+  desc.setUnknown();
+  descriptions.addDefault(desc);
+}
+
+int HiPuRhoProducer::ieta(const reco::CandidatePtr& in) const {
+  const CaloTower* ctc = dynamic_cast<const CaloTower*>(in.get());
+  if (ctc) {
+    return ctc->id().ieta();
+  }
+
+  throw cms::Exception("Invalid Constituent") << "CaloJet constituent is not of CaloTower type";
+
+  return 0;
+}
+
+int HiPuRhoProducer::iphi(const reco::CandidatePtr& in) const {
+  const CaloTower* ctc = dynamic_cast<const CaloTower*>(in.get());
+  if (ctc) {
+    return ctc->id().iphi();
+  }
+
+  throw cms::Exception("Invalid Constituent") << "CaloJet constituent is not of CaloTower type";
+
+  return 0;
+}
```

Comment:
not needed if CaloTower is used directly (it seems like there is no other alternative)